pilot/pkg/model/jwks_resolver.go

```diff
@@ -302,42 +316,45 @@ func (r *jwksResolver) refresh(t time.Time) {
 		jwksURI := key.(string)
 		e := value.(jwtPubKeyEntry)
 
-		// Remove cached item if it hasn't been used for a while.
-		if now.Sub(e.lastUsedTime) >= r.evictionDuration {
+		// Remove cached item for either of the following 2 situations
+		// 1) it hasn't been used for a while
+		// 2) it hasn't been refreshed successfully for a while
+		// This makes sure 2 things, we don't grow the cache infinitely and also we don't reuse a cached public key
+		// with no success refresh for too much time.
+		if now.Sub(e.lastUsedTime) >= r.evictionDuration || now.Sub(e.lastRefreshedTime) >= r.evictionDuration {
+			log.Infof("Removed cached JWT public key (lastRefreshed: %s, lastUsed: %s) from %q",
+				e.lastRefreshedTime, e.lastUsedTime, jwksURI)
 			r.keyEntries.Delete(jwksURI)
 			return true
 		}
 
 		oldPubKey := e.pubKey
 
-		// key rotation: fetch JWT public key again if it's expired.
-		if e.expireTime.Before(t) {
-			// Increment the WaitGroup counter.
-			wg.Add(1)
+		// Increment the WaitGroup counter.
+		wg.Add(1)
 
-			go func() {
-				// Decrement the counter when the goroutine completes.
-				defer wg.Done()
-
-				resp, err := r.getRemoteContent(jwksURI)
-				if err != nil {
-					log.Errorf("Cannot fetch JWT public key from %q, %v", jwksURI, err)
-					r.keyEntries.Delete(jwksURI)
-					return
-				}
-				newPubKey := string(resp)
+		go func() {
+			// Decrement the counter when the goroutine completes.
+			defer wg.Done()
 
-				r.keyEntries.Store(jwksURI, jwtPubKeyEntry{
-					pubKey:       newPubKey,
-					expireTime:   now.Add(r.expireDuration), // Update expireTime even if prev/current keys are the same.
-					lastUsedTime: e.lastUsedTime,            // keep original lastUsedTime.
-				})
-
-				if oldPubKey != newPubKey {
-					hasChange = true
-				}
-			}()
-		}
+			// Retry more aggressively in the refresh job.
+			resp, err := r.getRemoteContentWithRetry(jwksURI, 3)
```

Comment:
Do not put numbers in the code directly. Define a constant instead.  Same in a few other places.

Reply:
Done